vm/systemSmartContracts/stakingWaitingList.go

```diff
@@ -0,0 +1,1170 @@
+package systemSmartContracts
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const waitingListHeadKey = "waitingList"
+const waitingElementPrefix = "w_"
+
+type waitingListReturnData struct {
+	blsKeys         [][]byte
+	stakedDataList  []*StakedDataV2_0
+	lastKey         []byte
+	afterLastJailed bool
+}
+
+func (s *stakingSC) processStake(blsKey []byte, registrationData *StakedDataV2_0, addFirst bool) error {
```

Comment:
make processStakeV1 and V2 - one with legacy only with v4 in mind. in v4 you need to set the registernonce - line31 and the last 2 lines - like 51 and 52.

Reply:
Good suggestion, thank you, applied it !